network_manager/src/network_manager.cpp

```diff
@@ -0,0 +1,155 @@
+    {
+        return Ok(RUNNING);
+    }
+
+    res network_manager::run()
+    {
+        return Ok(RUNNING);
+    }
+
+    res network_manager::stop()
+    {
+        auto sta_ret = sta.stop();
+        auto ap_ret = ap.stop();
+        if (sta_ret.isErr() && ap_ret.isErr())
+        {
+            auto string = etl::make_string_with_capacity<128>("Error stopping Wifi. STA: ");
+            string += sta_ret.unwrapErr();
+            string += etl::make_string(" AP: ");
+            string += ap_ret.unwrapErr();
+            return Err(string);
+        }
+        else if (sta_ret.isErr())
+        {
+            auto string = etl::make_string_with_capacity<128>("Error stopping STA: ");
+            string += sta_ret.unwrapErr();
+            return Err(string);
+        }
+        else if (ap_ret.isErr())
+        {
+            auto string = etl::make_string_with_capacity<128>("Error stopping AP: ");
+            string += ap_ret.unwrapErr();
+            return Err(string);
+        }
+        return Ok(STOPPED);
+    }
+
+    res network_manager::initialize()
+    {
+        return Ok(RUNNING);
+    }
+
+    res network_manager::set_ap(bool state)
+    {
+        if (state)
+            return ap.initialize();
+        else
+            return ap.stop();
+    }
+
+    res network_manager::set_ap(bool state, wifi::ap_config_t conf)
+    {
+        m_config.ap = conf;
+        ap.set_config(conf);
+        return set_ap(state);
+    }
+
+    res network_manager::set_sta(bool state)
+    {
+        if (state)
+        {
+            auto ret = sta.initialize();
+            if (ret.isOk())
+            {
+                return set_ip_mode(m_config.dhcp_enable);
+            }
+            return ret;
+        }
+        else
+            return sta.stop();
+    }
+
+    res network_manager::set_sta(bool state, wifi::sta_config_t conf)
+    {
+        m_config.sta = conf;
+        sta.set_config(conf);
+        return set_sta(state);
+    }
+
+    res network_manager::set_ip_mode(bool state)
+    {
+        esp_err_t ret;
+        // State == true means DHCP mode
+        if (state)
+        {
+            ret = esp_netif_dhcpc_start(sta.get_netif());
+            if (ret != ESP_OK && ret != ESP_ERR_ESP_NETIF_DHCP_ALREADY_STARTED)
+                RETURN_ERR(ret);
+            else
+                return Ok(RUNNING);
+        }
+        else
+        {
+            ret = esp_netif_dhcpc_stop(sta.get_netif());
+            if (ret != ESP_OK && ret != ESP_ERR_ESP_NETIF_DHCP_ALREADY_STOPPED)
+                RETURN_ERR(ret);
+            else
+            {
+                esp_netif_ip_info_t ip4_conf;
+                inet_pton(AF_INET, m_config.ipv4_address.c_str(), &ip4_conf.ip);
+                inet_pton(AF_INET, m_config.ipv4_gateway.c_str(), &ip4_conf.gw);
+                inet_pton(AF_INET, m_config.ipv4_netmask.c_str(), &ip4_conf.netmask);
+
+                RETURN_ERR(esp_netif_set_ip_info(sta.get_netif(), &ip4_conf));
```

Comment:
Rename to `RETURN_ON_ERR`, now it's like the flow ends here regardless of what happens